### ISSUE TITLE
[IDEA] Provisioning of clients, domains and vhosts with/out shared DB/PHP-FPM

### ISSUE THREAD

**Issue by zenny** [msimerson/Mail-Toaster-6#447]
Hi @msimerson ,

I dared to ask this question because your replies are always swift, specific, and comprehensive. :+1: 

I do admire the way that you have separated the data and configs from the os/jails making it easy to migrate if needed. For that reason, I have been asking several developers (recent one in https://github.com/BastilleBSD/bastille/issues/234) to follow your suit.

However, I am trying to figure out whether MT6 can be used to generate multiple instances of jails for vhosts (like with ispconfig3 where each sshless user is created and vhosts are symbolically linked to the vhosts directory  or rudimentary freebsd script like http://microsux.dk/?p=204)  and/or dedicated hosts which shares some of the jails like db/php processes among others.

Fyi, if you are not aware of the ispconfig3 (I am sure you do), it creates several services with a single owner (like client1-N which owns services like http web1-N respectively). The clients are created without ssh access (_see example below_). That appears to me nifty in case of vhosts.

When I have gone through generated `mail-toaster.conf` and scripts under `provision` directory, it seems to me (pardon me if it went above my head) that they are hard-coded for a single owner (client) and single domains (web/mail) including services like haproxy configs. 

Won't it be a good idea to modify the script to create MT6 jail microservices per client basis with vhosts for subdomains with an option whether to share DB and PHP-FPM processes (the latter to reduce resource usage, similar to http://microsux.dk/wp-uploads/2010/10/setup.jpg)?

I meant to modify or append a script to create with this topology in view of affordable multi-core CPUs with cheap storage and memory devices which can house multiple tenants in a single machine (some even equipped with power- and mobo-redudancy):

```
Internet > Haproxy in load balancer setup > multiple web and mail servers either in concurrent or redundant state per client per domain > either dedicated/shared jails for DB (master-master handled by haproxy) and other process sharing like PHP-FPM
```

where,

1. The `mail-toaster.sh init` first creates at least a pair each of  load-balanced haproxy which also serves as centralized letsencrypt repo for all servers at the backend and DB jails system-wide.
2. Thereafter, MT6-script  shall create necessary services in separate jails with specific userspace permissions as stated below in `file-structure` based on FQDN-domain specified in `mail-toaster-FQDN.conf` (instead of using a generalized config name).

For example, let us say we have four T/CLDs (top/country level domains): domainA.tld, domainB.tld, domainC.cld, domainD.cld. domainA.tld and domainC.cld are owned by client1; and domainB.tld and domainD.cld are owned by client2. 

Although all clients and domains shares the same jails for DB (jailDB) and php-fpm(jailFPM) with their own permissions and authentication (dbname-password combo), the webservers and mailservers are hosted in different container jails under container directory viz A..D-jails where A- and C-jails only lets client1 access (permission) to all domains and subdomains underneath, and B- and D-jails only lets client2 access (permission) via FQDN-specific ACLs (by creating client1-N group with web1-N and mail1-N users). Further, the permission level should be set so specific that the users in A..D-jails cannot access each other's `config` and `data`. 

```
FILE-STRUCTURE/ 
├── config 
│   ├── client1 
│   │   ├── domainA.tld 
│   │   │   ├── dbA 
│   │   │   ├── mailA 
│   │   │   └── webA 
│   │   └── domainC.cld 
│   │       ├── dbC 
│   │       ├── mailC 
│   │       └── webC 
│   └── client2 
│       ├── domainB.tld 
│       │   ├── dbB 
│       │   ├── mailB 
│       │   └── webB 
│       └── domainD.cld 
│           ├── dbD 
│           ├── mailD 
│           └── webD 
├── jails 
│   ├── client1 
│   │   ├── domainA.tld 
│   │   │   ├── dbA 
│   │   │   ├── mailA 
│   │   │   └── webA 
│   │   └── domainC.cld 
│   │       ├── dbC 
│   │       ├── mailC 
│   │       └── webC 
│   └── client2 
│       ├── domainB.tld 
│       │   ├── dbB 
│       │   ├── mailB 
│       │   └── webB 
│       └── domainD.cld 
│           ├── dbD 
│           ├── mailD 
│           └── webD 
└── data 
    ├── client1 
    │   ├── domainA.tld 
    │   │   ├── dbA 
    │   │   ├── mailA 
    │   │   └── webA 
    │   └── domainC.cld 
    │       ├── dbC 
    │       ├── mailC 
    │       └── webC 
    └── client2
        ├── domainB.tld 
        │   ├── dbB 
        │   ├── mailB 
        │   └── webB 
        └── domainD.cld 
            ├── dbD 
            ├── mailD 
            └── webD 
57 directories, 0 files 


NB: 
1. The `config` part may be a sub-directory of `data` directory (default in MT6).
```

File permissions that may be useful in case of multiple vhosts in the same jail with per vhost userIDs can look like:

```
# ls -la ./config/clientA/domainA.tld/mailA/
total 1
drwxr-sr-x 2 mail1 client1 2 Aug  2 09:34 .

# ls -la ./config/clientA/domainA.tld/webA/
total 1
drwxr-sr-x 2 web1  client1 2 Aug  2 09:34 .

# ls -la ./config/clientA/domainA.tld/dbA/
total 1
drwxr-sr-x 2 data1  client1 2 Aug  2 09:34 .

# ls -la ./data/clientB/domainB.tld/mailB/
total 1
drwxr-sr-x 2 mail2  client2 2 Aug  2 09:34 .

# ls -la ./data/clientB/domainB.tld/webB/
total 1
drwxr-sr-x 2 web2  client2 2 Aug  2 09:34 .

# ls -la ./data/clientB/domainB.tld/dbB/
total 1
drwxr-sr-x 2 data2  client2 2 Aug  2 09:34 .

# ls -la ./jails/clientA/domainC.cld/webC/
total 1
drwxr-sr-x 2 web3  client1 2 Aug  2 09:34 .

# ls -la ./jails/clientB/domainD.cld/webD/
total 1
drwxr-sr-x 2 web4  client2 2 Aug  2 09:34 .

where `client1-N` replaces`clientA-Z` (due to useradd issue) and `mail1-N` replaces 'mailA-Z` and so on.
```

I believe something like this would make MT6 a gem to orchestrate FreeBSD jails in microservice architecture. 

Cheers, and stay safe,
/z

**Comment by msimerson** [comment]
### Short answer

Nah.

### TL;DR

It's not that I disagree with the premise, I have vhosts on my own servers that are almost exactly as you describe. However, I try hard to avoid those because they are inevitable maintenance problems. It's no coincidence that every web server has mass vhost configuration settings. It's by far the most efficient way to serve web pages and no matter how good your deployment tools are, it's going to be easier to comprehensively manage the config for fewer web servers than more of them.

**Comment by msimerson** [comment]
If I were looking to deploy many thousands or millions of vhosts today, there'd obviously be tiered products. The cheapest ones have to be on shared vhosts or the economics aren't competitive. At the other of the price spectrum, deploy each client to their own VM where they have strong isolation and guaranteed/enforced resource limits. Then you can provision each VM with exactly the services the client wants. If it's FreeBSD based, you can use MT6 as provisioning recipes, reminiscent of a cross between cPanel and a config management utility like chef or ansible.

These are cheap random thoughts, because I haven't given that space much consideration for many years. The last time I was seriously involved in deploying mass vhosts, I wrote [Provision::Unix](https://metacpan.org/release/Provision-Unix) as middleware between the physical hosts and a custom provisioning system that hooked into support tools, the client control panels, and the billing software. It had to interact with crappy hosting solutions like cPanel at the low end on up to several hypervisors running on Linux. It worked well enough to abstract away most of the provisioning work and focus on scaling up the business. Then we sold the business. 🤷‍♂️
